File: proto/move.js
function isArrInArr(a, b) {
  let result = false;
  b.forEach((arr) => {
    if (arr[0] === a[0] && arr[1] === a[1]) {
      result = true;
    }
  });
  return result;
}

function neighbors(x, y, grid) {
  const left = x > 0 ? [x - 1, y] : null;
  const right = x < grid[0].length - 1 ? [x + 1, y] : null;
  const up = y > 0 ? [x, y - 1] : null;
  const down = y < grid.length - 1 ? [x, y + 1] : null;
  const result = [];
  [left, right, up, down].forEach((neighbor) => {
    if (neighbor) {
      result.push(neighbor);
    }
  });
  return result;
}

class Path {
  constructor(history, grid) {
    this.history = history;
    this.moveCost = 0;
    history.slice(1).forEach((point) => {
      this.moveCost += grid[point[0]][point[1]];
    });
  }
}

function findPaths(unit, grid) {
  const paths = [];
  const coordinates = [unit.row, unit.col];
  let basePaths = [new Path([coordinates], grid)];
  for (let i = 0; i < 3; i += 1) {
    const newBasePaths = [];
    basePaths.forEach((path) => {
      neighbors(...path.history[path.history.length - 1], grid).forEach((n) => {
        if (!isArrInArr(n, path.history)) {
          const newPath = new Path([...path.history, n], grid);
          if (newPath.moveCost < unit.moveRange) {
            newBasePaths.push(newPath);
            paths.push(newPath);
          } else if (newPath.moveCost === unit.moveRange) {
            paths.push(newPath);
          }
        }
      });
    });
    basePaths = newBasePaths;
    paths.concat(basePaths);
  }
  return paths;
}

const testUnit = {
  moveRange: 2,
  row: 0,
  col: 0,
};
const testGrid = [
    [1, 1, 1, 1],
    [1, 2, 2, 1],
    [2, 1, 1, 1],
];
findPaths(testUnit, testGrid).forEach((path, num) => {
  console.log(`Path ${num}: `);
  path.history.forEach((point) => {
    console.log(point);
  });
});

export default findPaths;

File: proto/move.py
grid = [
    [1, 1, 1, 1, 2],
    [1, 2, 2, 1, 2],
    [2, 1, 1, 1, 2],
    [1, 1, 1, 2, 2],
    [1, 1, 1, 1, 2],
]

def neighbors(x, y):
    left = (x-1, y) if x > 0 else None
    right = (x+1, y) if x < len(grid[0])-1 else None
    up = (x, y-1) if y > 0 else None
    down = (x, y+1) if y < len(grid)-1 else None
    return [n for n in (left, right, up, down) if n]

class Path(object):
    def __init__(self, history):
        self.history = history
        self.move_cost = sum(grid[x][y] for x, y in history[1:])
    def __getitem__(self, index):
        return self.history[index]
    def __repr__(self):
        return str(self.history)
    # def step(new_point):
    #     self.history.append(new_point)
    #     self.move_cost += grid[new_point[0]][new_point[1]]

move_range = 2
position = (0, 0)
paths = []
base_paths = [Path([position])]

for i in range(move_range):
    new_base_paths = []
    for path in base_paths:
        for n in neighbors(*path[-1]):
            if n not in path:
                new_path = Path(path.history + [n])
                if new_path.move_cost < move_range:
                    new_base_paths.append(new_path)
                elif new_path.move_cost == move_range:
                    paths.append(new_path)
    paths.extend(base_paths)
    base_paths = new_base_paths

print(paths)
